### PR TITLE
Lucas/rust linter 1

### PR DESCRIPTION
**Pull request by franziskuskiefer**

### CODE REVIEW

**Comment by franziskuskiefer** [comment]
@W95Psp hm, looks like this isn't going through? I like the general ideas though.

**Comment by W95Psp** [comment]
Aahh, yes, sorry. I didn't try cargo clean before running cargo test.
The issue is that `cargo test` doesn't ensure the two binaries `cargo-circus` and the driver are built. We need to make dependencies explicit. https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#artifact-dependencies is the feature we want, but this is still not in stable cargo yet...

So, for now, I guess the best thing would be to run `cargo build --workspace --bins` before running any tests.

File: frontend/lint/tests/util.rs [hunk 1]
```diff
@@ -7,18 +8,31 @@ pub struct Test {
...
+        }
+        let mut cmd = Command::new(CARGO_CIRCUS.clone());
+        cmd.env("CIRCUS_RUSTC_DRIVER_BINARY", DRIVER.clone());
+        cmd.args(args);
+        // As documented in
+        // https://doc.rust-lang.org/cargo/reference/environment-variables.html#dynamic-library-paths,
+        // [cargo run] (and thus also [cargo test]) sets dynamic
+        // library paths, which causes some issues with dependencies
+        // when compiling without rustup
+        for env in ["DYLD_FALLBACK_LIBRARY_PATH", "LD_LIBRARY_PATH"] {
```
[the remark this answers is not in the document]
Comment (W95Psp) [reply frontend/lint/tests/util.rs:31]: `cargo run` sets them to specific values anyway, so (I think) manually inserting values here will be erased by Cargo anyway

If we don't remove this, building crate tests without rustup results in errors locating libraries. (and that's pretty un-debuggable, I had to replay the exact command with it's exact environment, and remove incrementally env vars to debug that...)

File: frontend/lint/tests/util.rs [hunk 2]
```diff
@@ -7,18 +8,31 @@ pub struct Test {
...
-    assert!(status.success());
+pub trait CommandCircusExt {
+    fn circus<I: IntoIterator<Item = S>, S: AsRef<OsStr>>(args: I) -> Self;
+}
+
+impl CommandCircusExt for Command {
+    fn circus<I: IntoIterator<Item = S>, S: AsRef<OsStr>>(args: I) -> Command {
+        use assert_cmd::cargo::cargo_bin;
+        use std::path::PathBuf;
+        lazy_static! {
```
[the remark this answers is not in the document]
Comment (W95Psp) [reply frontend/lint/tests/util.rs:19]: Well, it's recommended by the crate https://docs.rs/assert_cmd/latest/assert_cmd/cargo/index.html, so I just followed their recommendations.
As the number of test will grow, that will save time.

Oh, and I think I can just run the `cargo build --workspace --bin` in there as well, that will make things very easy.

**Comment by W95Psp** [comment]
I fixed things, @franziskuskiefer, does that look good to you?

File: frontend/lint/tests/util.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply frontend/lint/tests/util.rs:45]: I think it's fine to require rustup, and say so. But lets get this in for now.

**Pull request merged**